In [1]:
import pandas as pd
import numpy as np

cleveland_data = pd.read_csv("processed.cleveland.data", header=None)
hungarian_data = pd.read_csv("processed.hungarian.data", header=None)
switzerland_data = pd.read_csv("processed.switzerland.data", header=None)
va_data = pd.read_csv("processed.va.data", header=None)

In [2]:
cleveland_data.head()

0    1    2      3      4    5    6      7    8    9    10   11   12  13
0  63.0  1.0  1.0  145.0  233.0  1.0  2.0  150.0  0.0  2.3  3.0  0.0  6.0   0
1  67.0  1.0  4.0  160.0  286.0  0.0  2.0  108.0  1.0  1.5  2.0  3.0  3.0   2
2  67.0  1.0  4.0  120.0  229.0  0.0  2.0  129.0  1.0  2.6  2.0  2.0  7.0   1
3  37.0  1.0  3.0  130.0  250.0  0.0  0.0  187.0  0.0  3.5  3.0  0.0  3.0   0
4  41.0  0.0  2.0  130.0  204.0  0.0  2.0  172.0  0.0  1.4  1.0  0.0  3.0   0

In [3]:
hungarian_data.head()

0   1   2    3    4  5  6    7  8    9  10 11 12  13
0  28   1   2  130  132  0  2  185  0  0.0  ?  ?  ?   0
1  29   1   2  120  243  0  0  160  0  0.0  ?  ?  ?   0
2  29   1   2  140    ?  0  0  170  0  0.0  ?  ?  ?   0
3  30   0   1  170  237  0  1  170  0  0.0  ?  ?  6   0
4  31   0   2  100  219  0  1  150  0  0.0  ?  ?  ?   0

In [4]:
switzerland_data.head()

0   1   2    3   4  5  6    7  8    9  10 11 12  13
0  32   1   1   95   0  ?  0  127  0   .7  1  ?  ?   1
1  34   1   4  115   0  ?  ?  154  0   .2  1  ?  ?   1
2  35   1   4    ?   0  ?  0  130  1    ?  ?  ?  7   3
3  36   1   4  110   0  ?  0  125  1    1  2  ?  6   1
4  38   0   4  105   0  ?  0  166  0  2.8  1  ?  ?   2

In [5]:
va_data.head()

0   1   2    3    4  5   6    7  8    9  10 11 12  13
0  63   1   4  140  260  0   1  112  1    3  2  ?  ?   2
1  44   1   4  130  209  0   1  127  0    0  ?  ?  ?   0
2  60   1   4  132  218  0   1  140  1  1.5  3  ?  ?   2
3  55   1   4  142  228  0   1  149  1  2.5  1  ?  ?   1
4  66   1   3  110  213  1   2   99  1  1.3  2  ?  ?   0

From the Heart Diseases dataset, we have data from 4 different areas, Cleveland, Hungary, Switzerland and Virginia. We want to combine these datasets together. We will first change the column names.

According to the attribute information, the following 14 attributes were used.

1.   age
2.   sex   
3.   cp   
4.   trestbps
5.   chol   
6.   fbs     
7.   restecg 
8.   thalach
9.   exang   
10.  oldpeak  
11.  slope   
12.  ca       
13.  thal     
14.  num (the predicted attribute)


Here is the description and type of each of the attributes:
      
1.   age: age in years (**Numerical**)

2.   sex: sex (1 = male; 0 = female) (**Categorical**)

3.   cp: chest pain type (**Categorical**)
       - Value 1: typical angina
       - Value 2: atypical angina
       - Value 3: non-anginal pain
       - Value 4: asymptomatic
    
4.   trestbps: resting blood pressure (in mm Hg on admission to the hospital) (**Numerical**)

5.   chol: serum cholestoral in mg/dl (**Numerical**)

6.   fbs: (fasting blood sugar > 120 mg/dl)  (1 = true; 0 = false) (**Categorical**)

7.   restecg: resting electrocardiographic results (**Categorical**)
       - Value 0: normal
       - Value 1: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)
       - Value 2: showing probable or definite left ventricular hypertrophy by Estes' criteria
                    
8.   thalach: maximum cleveland rate achieved (**Numerical**)

9.   exang: exercise induced angina (1 = yes; 0 = no) (**Categorical**)

10.  oldpeak = ST depression induced by exercise relative to rest (**Numerical**)

11.  slope: the slope of the peak exercise ST segment (**Categorical**)
       - Value 1: upsloping
       - Value 2: flat
       - Value 3: downsloping
        
12.  ca: number of major vessels (0-3) colored by flourosopy (**Numerical**)

13.  thal: 3 = normal; 6 = fixed defect; 7 = reversable defect (**Categorical**)

14.  num: diagnosis of cleveland disease (angiographic disease status) (**Categorical**)
       - Value 0: < 50% diameter narrowing
       - Value 1: > 50% diameter narrowing
        

In [6]:
# Creating a copy of the datasets
cleveland_clean = cleveland_data.copy()
hungarian_clean = hungarian_data.copy()
switzerland_clean = switzerland_data.copy()
va_clean = va_data.copy()

In [7]:
# Adding column titles to all the datasets
cleveland_clean.columns = ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num']
hungarian_clean.columns = ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num']
switzerland_clean.columns = ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num']
va_clean.columns = ['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalch', 'exang', 'oldpeak', 'slope', 'ca', 'thal', 'num']

In [8]:
# Convert all Variable Names to UPPERCASE
cleveland_clean.columns = cleveland_clean.columns.str.upper()
hungarian_clean.columns = hungarian_clean.columns.str.upper()
switzerland_clean.columns = switzerland_clean.columns.str.upper()
va_clean.columns = va_clean.columns.str.upper()

We will add a new column with the location. The following are the short forms used:
   * CL - Cleveland 
   * HU - Hungary
   * SW - Switzerland
   * VA - Virginia

In [9]:
cleveland_clean['LOC'] = 'CL'
hungarian_clean['LOC'] = 'HU'
switzerland_clean['LOC'] = 'SW'
va_clean['LOC'] = 'VA'

In [10]:
# Print the Variable Information to check
cleveland_clean.info()
hungarian_clean.info()
switzerland_clean.info()
va_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 15 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   AGE       303 non-null    float64
 1   SEX       303 non-null    float64
 2   CP        303 non-null    float64
 3   TRESTBPS  303 non-null    float64
 4   CHOL      303 non-null    float64
 5   FBS       303 non-null    float64
 6   RESTECG   303 non-null    float64
 7   THALCH    303 non-null    float64
 8   EXANG     303 non-null    float64
 9   OLDPEAK   303 non-null    float64
 10  SLOPE     303 non-null    float64
 11  CA        303 non-null    object 
 12  THAL      303 non-null    object 
 13  NUM       303 non-null    int64  
 14  LOC       303 non-null    object 
dtypes: float64(11), int64(1), object(3)
memory usage: 35.6+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 294 entries, 0 to 293
Data columns (total 15 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    ---------

We notice that there are ? in the dataset. We shall change these to NULL.

In [11]:
# Changing ? in dataframe to NaN
cleveland_clean = cleveland_clean.replace('?', np.nan)
hungarian_clean = hungarian_clean.replace('?', np.nan)
switzerland_clean = switzerland_clean.replace('?', np.nan)
va_clean = va_clean.replace('?', np.nan)

In [12]:
cleveland_clean.isna().sum()

AGE         0
SEX         0
CP          0
TRESTBPS    0
CHOL        0
FBS         0
RESTECG     0
THALCH      0
EXANG       0
OLDPEAK     0
SLOPE       0
CA          4
THAL        2
NUM         0
LOC         0
dtype: int64

In [13]:
hungarian_clean.isna().sum()

AGE           0
SEX           0
CP            0
TRESTBPS      1
CHOL         23
FBS           8
RESTECG       1
THALCH        1
EXANG         1
OLDPEAK       0
SLOPE       190
CA          291
THAL        266
NUM           0
LOC           0
dtype: int64

In [14]:
switzerland_clean.isna().sum()

AGE           0
SEX           0
CP            0
TRESTBPS      2
CHOL          0
FBS          75
RESTECG       1
THALCH        1
EXANG         1
OLDPEAK       6
SLOPE        17
CA          118
THAL         52
NUM           0
LOC           0
dtype: int64

In [15]:
va_clean.isna().sum()

AGE           0
SEX           0
CP            0
TRESTBPS     56
CHOL          7
FBS           7
RESTECG       0
THALCH       53
EXANG        53
OLDPEAK      56
SLOPE       102
CA          198
THAL        166
NUM           0
LOC           0
dtype: int64

We notice that there are a lot of missing values in the datasets for Hungary, Switzerland and Virginia. If we were to drop the columns or rows, this would result in a lot of missing data and limits our analysis. Therefore, we shall stick with the Cleveland dataset since it only has 6 rows with missing information.

In [16]:
# Dropping 6 rows with NaN values.
cleveland_clean = cleveland_clean.dropna()

In [17]:
# Printing the Variable Information to check after deleting NaN rows
cleveland_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 297 entries, 0 to 301
Data columns (total 15 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   AGE       297 non-null    float64
 1   SEX       297 non-null    float64
 2   CP        297 non-null    float64
 3   TRESTBPS  297 non-null    float64
 4   CHOL      297 non-null    float64
 5   FBS       297 non-null    float64
 6   RESTECG   297 non-null    float64
 7   THALCH    297 non-null    float64
 8   EXANG     297 non-null    float64
 9   OLDPEAK   297 non-null    float64
 10  SLOPE     297 non-null    float64
 11  CA        297 non-null    object 
 12  THAL      297 non-null    object 
 13  NUM       297 non-null    int64  
 14  LOC       297 non-null    object 
dtypes: float64(11), int64(1), object(3)
memory usage: 37.1+ KB


We noticed that for the NUM attribute, there were numbers ranging from 0 to 4, however we decided to only have 2 categories. The first category would be presence of heart disease, represented by 1 and the second category would be absence of heart disease, represented by 0.

In [18]:
cleveland_clean['DIS'] = np.where(cleveland_clean['NUM']==0, 0, 1)

In [19]:
cleveland_clean.to_csv('clean_heart_data.csv')